docs/proposals/network-policy.md

```diff
@@ -0,0 +1,136 @@
+# Native API Support for Network Policy
+#### Abstract
+A proposal for adding the API primitives to describe flexible network policy.
+
+#### Goals
+- Add first-class support for flexible network policy to the API.
+- Provide an intent-based API which uses native Kubernetes concepts (labels, namespaces, services).
+- Abstracts policy from the underlying networking technology.
+- Back-compatible with existing network plugins.  Network plugins can choose which (if any) of this API is supported.
+
+#### Proposed Design
+The design uses Kubernetes labels to apply policy to pods.  A NetworkPolicy object is introduced which uses a label
+selector to apply a list of intent-based rules to groups of pods.  If traffic matches any rule within the given list, it
+should be allowed to the destination pod.  If no rules are matched, it should be rejected.
+
+Each rule in the NetworkPolicy can apply policy on any combination of the following fields:
+- Kubernetes Namespace
+- Pod Labels
+- Protocol (TCP vs UDP)
+- Target port
+
+The rules can be extended to support additional fields in the future.
+
+##### New API Objects
+I'd like to propose the following set of new first-class API objects
+```
+type NamespaceGroup string
+
+const (
+  // Represents all namespaces.
+  NamespaceGroupAll NamespaceGroup = "AllNamespaces"
+)
+
+// A rule which can be applied to a pod or groupd of pods
+// using a NetworkPolicy.
+type NetworkPolicyRule struct {
+  // Allow from / to this namespace.  Accepts any existing namespace,
+  // or the special "AllNamespaces" value which selects all namespaces.
+  Namespace string `json:"selector,omitempty"`
+  
+  // Allow from / to these labels.
+  Labels map[string]string `json:"labels,omitempty"`
+  
+  // Allow on this protocol.
+  Protocol string `json:"protocols,omitempty"`
+ 
+  // Allow to this port on the given protocol.
+  ToPort int `json:"ports,omitempty"`
+}
+
+// Specification for a NetworkPolicy object.
+type NetworkPolicySpec struct {
+        // Selector - determines which pods will have this NetworkPolicy applied.
+        // If empty, all pods are selected, if not specified, endpoints must be manually specified.
+        Selector map[string]string `json:"selector,omitempty"`
+        
+        // List of rules to apply to pods selected by this NetworkPolicy.
+        // These rules specify which traffic is allowed by this policy.
+        Rules [ ]NetworkPolicyRule `json:rules,omitempty"`
+}
+
+// NetworkPolicy includes a label selector to pick pods, and a list of network policy rules 
+// which should be applied to the selected pods.
+type NetworkPolicy struct {
+        // Standard object's metadata.
+        ObjectMeta `json:"metadata,omitempty"`
+        
+        // Specification - includes selector and rules.
+        Spec NetworkPolicySpec `json:"spec,omitempty"`
```

Comment:
Hm.

I was not planning on having the policy defined on the namespace object itself (rather it would just create a corresponding NetworkPolicy object) but maybe that is the right way around.